src/js/actions/libraries.js

```diff
@@ -0,0 +1,272 @@
+        libraryAdapter = require("adapter/lib/libraries");
+
+    var events = require("../events"),
+        locks = require("../locks");
+
+    /**
+     * Uploads the selected single layer to the current library
+     *
+     * Achieves this by:
+     *  - Creates a new element in the library
+     *  - Calls PS to export the layer to a temporary location
+     *  - Passes the temporary path to libraries API to update layer's content
+     *  - Tells Photoshop the location of the content
+     *  - Updates the document
+     *
+     * Eventually, we'll need this to accept layer(s), library, and be more flexible
+     * Also, we definitely need to get rid of the update document call, but this is 0.1
+     *
+     * @return {Promise}
+     */
+    var createElementFromSelectedLayer = function () {
+        var appStore = this.flux.store("application"),
+            libStore = this.flux.store("library"),
+            currentDocument = appStore.getCurrentDocument(),
+            currentLibrary = libStore.getCurrentLibrary(),
+            currentLayers = currentDocument.layers.selected;
+
+        if (!currentLibrary || currentLayers.count() !== 1) {
+            return Promise.resolve();
+        }
+
+        var currentLayer = currentLayers.first(),
+            IMAGE_ELEMENT_TYPE = "application/vnd.adobe.element.image+dcx",
+            REPRESENTATION_TYPE = "image/vnd.adobe.photoshop";
+
+        currentLibrary.beginOperation();
+
+        var newElement = currentLibrary.createElement(currentLayer.name, IMAGE_ELEMENT_TYPE),
+            representation = newElement.createRepresentation(REPRESENTATION_TYPE, "primary"),
+            previewSize = {
+                w: 248,
+                h: 188
+            };
+
+        // FIXME: Mac/Win temporary locations!
+        var exportObj = libraryAdapter.exportLayer("/tmp/", "/tmp/preview.png",
+            currentLayer.name, previewSize);
+
+        return descriptor.playObject(exportObj)
+            .bind(this)
+            .then(function (saveData) {
+                var path = saveData.in._path;
+
+                return Promise.fromNode(function (cb) {
+                    representation.updateContentFromPath(path, false, cb);
+                });
+            }).finally(function () {
+                currentLibrary.endOperation();
+            }).then(function () {
+                var newRepresentation = newElement.getPrimaryRepresentation();
+                return Promise.fromNode(function (cb) {
+                    newRepresentation.getContentPath(cb);
+                });
+            }).then(function (path) {
+                var createObj = libraryAdapter.createElement(currentDocument.id, currentLayer.id, newElement, path);
+                return descriptor.playObject(createObj);
+            }).then(function () {
+                // FIXME: Find a way around this update Document
+                this.flux.actions.documents.updateDocument();
```

Comment:
Can you make a core Trello card for all the instances in which we need some kind of layer ID in the action descriptor result, or whatever it takes to get rid of the updateDocument calls? If you do that then we can try to find someone else on the core team to add those API features in parallel.